feat: create single Nomad server running on DigitalOcean droplet

cdunster opened the pull request:
This is the Pulumi definition for the Nomad Server that is required for #holochain/wind-tunnel/issues/132 and can be found at https://nomad-server-01.holochain.org:4646/ui.

cdunster committed:
feat: add hashicorp apt repo for latest nomad version

cdunster committed:
feat: write nomad config to default location

cdunster committed:
feat: start nomad as a service

cdunster committed:
feat: run nomad server as nomad user

cdunster committed:
feat: enable ACL to restrict access to Nomad server

cdunster committed:
feat: advertise public IPv4 address on server

cdunster committed:
feat: create reserved IPv4 for nomad server

cdunster committed:
chore: add nomad package

cdunster committed:
chore: add nomad client config for testing

cdunster committed:
feat: untie reserved IP from droplet lifecycle

Before this, changing the droplet replaced the reserved IP address which
changed the actual address.

cdunster committed:
refactor: don't unnecessarily convert droplet size to string

cdunster committed:
feat: enable TLS and create certs on boot

cdunster committed:
feat: don't advertise serf via public IP

cdunster committed:
docs: add section in README for instructions to deploy a server

cdunster committed:
feat: allow clients to connect without TLS

cdunster committed:
feat: add nomad-server-01.holochain.org as DNS name for certificate

cdunster committed:
feat: remove broken verifying server hostname option

It's not working so just remove it for now.

cdunster committed:
feat: increase job lifetime to 24 hours

holochain-release-automation2 commented:
#### :tropical_drink: `preview` on nomad-server/holochain/nomad-server

  <details>
  <summary>Pulumi report</summary>
  
[View in Pulumi Cloud](https://app.pulumi.com/holochain/nomad-server/nomad-server/previews/b1f3a3ca-5570-488c-8047-d17d0b61afac)

  
  <pre>
  Previewing update (holochain/nomad-server)

View Live: https://app.pulumi.com/holochain/nomad-server/nomad-server/previews/b1f3a3ca-5570-488c-8047-d17d0b61afac

pulumi:pulumi:Stack: (same)
  [urn=urn:pulumi:nomad-server::nomad-server::pulumi:pulumi:Stack::nomad-server-nomad-server]
  +-command:remote:Command: (replace)
      [id=chown-etc-nomad-dir-before-server-certc71d31ab]
      [urn=urn:pulumi:nomad-server::nomad-server::command:remote:Command::chown-etc-nomad-dir-before-server-cert]
      [provider=urn:pulumi:nomad-server::nomad-server::pulumi:providers:command::default_1_0_2::3d5f4907-5ccd-4d31-8098-9549b40692ec]
    - triggers: [secret]
    + triggers: [
    +     [0]: asset(file:f687a96) { ./nomad-agent-ca.pem }
    +     [1]: [secret]
    +     [2]: asset(file:6775693) { ./nomad.hcl }
      ]
  +-command:remote:Command: (replace)
      [id=create-server-cert1ba48512]
      [urn=urn:pulumi:nomad-server::nomad-server::command:remote:Command::create-server-cert]
      [provider=urn:pulumi:nomad-server::nomad-server::pulumi:providers:command::default_1_0_2::3d5f4907-5ccd-4d31-8098-9549b40692ec]
    - triggers: [secret]
    + triggers: [
    +     [0]: asset(file:f687a96) { ./nomad-agent-ca.pem }
    +     [1]: [secret]
      ]
  +-command:remote:Command: (replace)
      [id=print-server-cert13df651b]
      [urn=urn:pulumi:nomad-server::nomad-server::command:remote:Command::print-server-cert]
      [provider=urn:pulumi:nomad-server::nomad-server::pulumi:providers:command::default_1_0_2::3d5f4907-5ccd-4d31-8098-9549b40692ec]
    - triggers: [secret]
    + triggers: [
    +     [0]: asset(file:f687a96) { ./nomad-agent-ca.pem }
    +     [1]: [secret]
      ]
  +-command:remote:Command: (replace)
      [id=chown-etc-nomad-dir52b48959]
      [urn=urn:pulumi:nomad-server::nomad-server::command:remote:Command::chown-etc-nomad-dir]
      [provider=urn:pulumi:nomad-server::nomad-server::pulumi:providers:command::default_1_0_2::3d5f4907-5ccd-4d31-8098-9549b40692ec]
    - triggers: [secret]
    + triggers: [
    +     [0]: asset(file:f687a96) { ./nomad-agent-ca.pem }
    +     [1]: [secret]
    +     [2]: asset(file:6775693) { ./nomad.hcl }
    +     [3]: asset(file:1289f09) { ./job-runner.policy.hcl }
      ]
  +-command:remote:Command: (replace)
      [id=apply-job-runner-policy8353b57a]
      [urn=urn:pulumi:nomad-server::nomad-server::command:remote:Command::apply-job-runner-policy]
      [provider=urn:pulumi:nomad-server::nomad-server::pulumi:providers:command::default_1_0_2::3d5f4907-5ccd-4d31-8098-9549b40692ec]
    - triggers: [secret]
    + triggers: [
    +     [0]: asset(file:1289f09) { ./job-runner.policy.hcl }
    +     [1]: [secret]
      ]
Resources:
  +-5 to replace
  17 unchanged
  </pre>
  
  </details>

ThetaSinner commented on main.go:
```diff
@@ -14,6 +16,14 @@ func main() {
 
 	pulumi.Run(func(ctx *pulumi.Context) error {
+		region := digitalocean.RegionAMS3
+		reservedIp, err := digitalocean.NewReservedIp(ctx, "nomad-server-01-reserved-ip", &digitalocean.ReservedIpArgs{
```
Why did the reserved IP end up being needed?

cdunster replied:
Because nomad-server-01.holochain.org points to this reserved IP. I think that we could manage the domain with DigitalOcean but I don't have the permissions to do that.

cdunster committed:
docs: fix typo in README

cdunster committed:
chore: update go packages

cdunster committed:
feat: add ssh key for this workflow to DigitalOcean

cdunster committed:
feat!: remove nomad setup via cloud-init

This will be handled by Pulumi SSH commands.

BREAKING CHANGE: Will re-create the droplet.

cdunster committed:
fix: make reserved IP assignment depend on droplet and reserved IP

cdunster committed:
feat: copy existing CA cert

cdunster committed:
feat: copy nomad.service file

cdunster committed:
feat: copy nomad.hcl agent config

cdunster committed:
feat: make sure /opt/nomad/data exists and is owned by nomad

Should be created when nomad is installed but let's make sure.

cdunster committed:
feat: enable and start the nomad service

cdunster committed:
feat: copy secret CA cert key

cdunster committed:
feat: create the server-specific cert

cdunster committed:
fix: avoid ssh command timeout

Set the max attempts to unlimited

cdunster committed:
fix: make sure nomad user exists before using

cdunster committed:
feat: bootstrap ACL with existing token

cdunster committed:
fix: wait for reserved IP to be assigned before SSH

Pulumi can't assign to droplet whilst SSH'ing

cdunster committed:
feat: add and apply job-runner policy

cdunster committed:
feat: add triggers and deps to re-run steps if things change

cdunster committed:
fix: delete old server cert when creating new one

cdunster committed:
fix: hide stdout when bootstrapping ACL

cdunster committed:
feat: reduce ssh timeout from infinite to 60

jost-s commented on README.md:
```diff
@@ -56,3 +56,35 @@ This value is encrypted by Pulumi and stored in [Pulumi.nomad-server.yaml].
+Some changes to the definition of the DigitalOcean droplet will require the
+droplet to be destroyed and re-created.
+
+If this happens or if we want to deliberately re-create the server then the TLS
```
```suggestion
If this happens or if we want to deliberately re-create the server, then the TLS
```

cdunster replied:
Done as part of: d6eb769286d3857517fec5fd137edcad288a91aa

jost-s commented on README.md:
```diff
@@ -56,3 +56,35 @@ This value is encrypted by Pulumi and stored in [Pulumi.nomad-server.yaml].
+bootstrap the ACL support and update the token in BitWarden used by developers
+to access and manage the server.
+
+First, SSH into the Nomad server, this should be accessible at
+`nomad-server-01.holochain.org` and if your SSH key is in DigitalOcean then you
+should also have access to is.
```
```suggestion
First SSH into the Nomad server. This should be accessible at
`nomad-server-01.holochain.org`, and if your SSH key is in DigitalOcean, then you
should also have access to it.
```

cdunster replied:
I actually removed this section as part of d6eb769286d3857517fec5fd137edcad288a91aa as now Pulumi performs these commands for us :slightly_smiling_face:

cdunster committed:
fix: use single nomad config file instead of dir

Nomad was trying to read the policy as a config.

cdunster committed:
fix: add dependency on creating SSH key for getting SSH keys

Sometimes the droplet was created without this access key that is needed
later.

cdunster committed:
feat: update the nomad service start command

cdunster committed:
chore: update all certs and keys after stack delete

cdunster committed:
fix: use bootstrap token to add job-runner policy

cdunster committed:
fix: correct Triggers as they don't work on previous jobs

mattyg commented on Pulumi.nomad-server.yaml:
```diff
@@ -1,3 +1,10 @@
+    secure: AAABAGXgpPfJ/xG/pzCArMjgdcJdwLaPMdqM2OFEEPy1cQjfkpHVRFE+BGgniLBpcuMsnIDxTEaVqEeU27IyeiS4mmA=
+  nomad-server:caCertKey:
+    secure: AAABALxHeyRQnIH7MMVTG2nmFV0VY7uBPU/ATstRp80+hecERGSxI2id93ymeH6DwsiL/4sgn2iRnaxQTYAFMzH0Tucb/ZQejM2Y85y3dEiiAo1N0mjEKps15W7SORZuOWH1xqnD8CRDysHzkSr59bEmvt2r1Wjccq0OFzCtCqyLD0D/fgLI3W/PcJSSE4dB8FsES5LAwiDVB/3xnUqh1NA3oKjL8jAN6xxbjH6Q9ohgg9eaGNBrvsZsNazpgL7Z4NApd0y0uq+XoDrIBskwixNT3Jn5Iw5Quynr7xU9cExtBenRtg/izEM1cI6Vit+z9NLeR/i/mbC+djMlbfqnKPMi
+  nomad-server:serverAccessPrivateKey:
```
are these secrets we should be keeping out of git?

cdunster replied:
Pulumi encrypts them with the idea that these files are stored in VSC so it should be safe.

cdunster committed:
chore: update config after moving stack

cdunster committed:
fix: copy policy file depends on etc/nomad.d dir

cdunster committed:
docs: fix README format issues

cdunster committed:
docs: update the README

cdunster commented:
@mattyg, @jost-s, and @ThetaSinner, I've added quite a lot to the Pulumi code since your last reviews. Now, Pulumi performs the Nomad setup for us automatically.

cdunster committed:
feat: add InfluxDB token to server

cdunster committed:
docs: add section about replacing the InfluxDB token

cdunster committed:
feat: add step to print the new server certificate

cdunster committed:
docs: fix typo in README

cdunster committed:
docs: add extra info about resetting the ACL token in README

jost-s commented on README.md:
```diff
@@ -44,15 +56,146 @@ changes to Pulumi.
+
+## Changing the TLS certificate authority certificate
+
+Nomad uses TLS for encrypted communications with the server(s) for this to work
```
```suggestion
Nomad uses TLS for encrypted communications with the server(s). For this to work
```

jost-s commented on README.md:
```diff
@@ -44,15 +56,146 @@ changes to Pulumi.
+Some changes to the definition of the DigitalOcean droplet will require the
+droplet to be destroyed and re-created.
+
+If this happens, or if we want to deliberately re-create the server(s), then
```
```suggestion
If this happens, or if we want to deliberately re-create the server(s),
```

jost-s reviewed (commented):
I'm too late with my review, but anyway. I wanted to run the commands today, so here is my text from yesterday.

You went to town on the go coding there!
I can follow the description of all the certificates and tokens. I trust that it works, I haven't run the commands myself.

cdunster commented:
Ah, sorry @jost-s. I thought that it seems to be working, and we can always make changes later in separate PRs, so I just went for it :grin:

> You went to town on the go coding there!

Yeah, once I found the command to run arbitrary commands and copy files via SSH then I saw no reason to not just automate everything

jost-s commented:
Totally agree. I'll come back to this if necessary once I've tried out these steps.
